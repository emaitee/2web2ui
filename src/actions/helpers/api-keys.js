import _ from 'lodash';
import { getGrants } from 'src/selectors/api-keys';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = _.keys(getGrants(getState()));
  } else {
    request.data.grants = _.keys(_.pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(_.trim);
  }

  return request;
}
