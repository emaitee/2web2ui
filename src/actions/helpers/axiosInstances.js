import axios from 'axios';
import config from 'src/config';

const { apiBase, apiRequestTimeout, zuora: zuoraConfig } = config;

export const sparkpost = axios.create({
  baseURL: apiBase,
  timeout: apiRequestTimeout
});

export const zuora = axios.create({
  baseURL: zuoraConfig.baseUrl,
  timeout: zuoraConfig.timeout
});
