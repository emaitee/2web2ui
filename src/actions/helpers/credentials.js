import { pickBy, trim, keys } from 'lodash';

import { getGrants } from 'selectors/credentials';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = keys(getGrants(getState()));
  } else {
    request.data.grants = keys(pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(trim);
  }

  return request;
}
