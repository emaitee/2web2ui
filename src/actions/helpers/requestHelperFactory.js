import axios from 'axios';

function defaultOnSuccess({ types, response, dispatch, meta, action }) {
  dispatch({
    type: types.SUCCESS,
    payload: response,
    meta
  });

  return response;
}

function defaultOnFail({ types, err, dispatch, meta, action }) {
  const { message, response = {}} = err;

  dispatch({
    type: types.FAIL,
    payload: { message, response },
    meta
  });
}

export default function requestFactory({
  request = axios,
  onSuccess = defaultOnSuccess,
  onFail = defaultOnFail,
  transformHttpOptions = (opts) => opts
} = {}) {
  return (action) => (dispatch, getState) => {
    const { type = 'NO_TYPE_DEFINED', meta } = action;
    const { url, method = 'get', params, headers, data } = meta;
    const types = {
      PENDING: `${type}_PENDING`,
      SUCCESS: `${type}_SUCCESS`,
      FAIL: `${type}_FAIL`
    };

    // standardize method included in meta from here on out
    meta.method = method.toLowerCase();

    dispatch({
      type: types.PENDING,
      meta
    });

    const httpOptions = {
      method: meta.method,
      url,
      params,
      headers,
      data
    };

    const transformedHttpOptions = transformHttpOptions(httpOptions, getState);

    return request(transformedHttpOptions)
      .then(
        // request succeeded, we only get here if the request returned a 2xx status code
        (response) => onSuccess({ types, response, dispatch, meta, action, getState }),

        // request failed (remember to throw err in your onFail)
        (err) => {
          onFail({ types, err, dispatch, meta, action, getState });
          return { error: err }
        }
      );
  };
}
