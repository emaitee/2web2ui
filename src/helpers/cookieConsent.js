import * as jsCookie from 'js-cookie';
import config from 'src/config';

const { name, ageDays, options } = config.cookieConsent.cookie;

const isCookieSet = () => jsCookie.get(name) !== undefined;

const setCookie = () => jsCookie.set(name, '', { ...options, expires: ageDays });

export default { isCookieSet, setCookie };
