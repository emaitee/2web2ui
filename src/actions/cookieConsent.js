import helpers from 'src/helpers/cookieConsent';

export const giveConsent = () => {
  helpers.setCookie();
  return { type: 'GIVE_COOKIE_CONSENT' };
};

