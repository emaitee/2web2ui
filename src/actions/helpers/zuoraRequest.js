import requestHelperFactory from './requestHelperFactory';
import _ from 'lodash';
import { zuora as zuoraAxios } from './axiosInstances';

export default requestHelperFactory({
  request: zuoraAxios,
  onSuccess: ({ types, response, dispatch, meta }) => {
    if (!response.data.success) {
      const message = _.get(response, 'data.reasons[0].message', 'An error occurred while contacting the billing service');

      dispatch({
        type: types.FAIL,
        payload: { message, response },
        meta
      });

      const error = new Error(message);
      error.response = response;

      return { error };
    }

    dispatch({
      type: types.SUCCESS,
      payload: response,
      meta
    });

    return response;
  }
});
