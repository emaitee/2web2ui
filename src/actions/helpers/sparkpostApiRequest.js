import requestHelperFactory from 'src/actions/helpers/requestHelperFactory';
import { refresh, logout } from 'src/actions/auth';
import { showAlert } from 'src/actions/globalAlert';
import { useRefreshToken } from 'src/helpers/http';
import { resolveOnCondition } from 'src/helpers/promise';
import _ from 'lodash';
import { sparkpost as sparkpostAxios } from './axiosInstances';

const maxRefreshRetries = 3;
export const refreshTokensUsed = new Set();
export let refreshing = false;

// Re-dispatches a given action after we finish refreshing the auth token
function redispatchAfterRefresh(action, dispatch) {
  return resolveOnCondition(() => !refreshing).then(() => dispatch(sparkpostRequest(action)));
}

const sparkpostRequest = requestHelperFactory({
  request: sparkpostAxios,
  transformHttpOptions: (options, getState) => {
    const { auth } = getState();
    const transformed = { ...options };
    if (auth.loggedIn) {
      _.set(transformed, 'headers.Authorization', auth.token);
    }
    return transformed;
  },
  onSuccess: ({ response, dispatch, types, meta }) => {
    const { data: { results }} = response;
    dispatch({
      type: types.SUCCESS,
      payload: results,
      meta
    });

    return results;
  },
  onFail: ({ types, err, dispatch, meta, action, getState }) => {
    const { message, response = {}} = err;
    const { auth } = getState();
    const { retries = 0 } = meta;

    // NOTE: if this is a 401 and we have a refresh token, we need to do a
    // refresh to get a new auth token and then re-dispatch this action
    if (response.status === 401 && auth.refreshToken && retries < maxRefreshRetries) {
      action.meta.retries = retries + 1;

      // If we are currently refreshing the token OR if this refresh token
      // has already been used to refresh, we should re-dispatch after refresh is complete
      if (refreshing || refreshTokensUsed.has(auth.refreshToken)) {
        return redispatchAfterRefresh(action, dispatch);
      }

      refreshing = true;
      refreshTokensUsed.add(auth.refreshToken);

      // call API for a new token
      return useRefreshToken(auth.refreshToken)

        // dispatch a refresh action to save new token results in cookie and store
        .then(({ data } = {}) => dispatch(refresh(data.access_token, data.refresh_token)))

        // dispatch the original action again, now that we have a new token ...
        // if anything in this refresh flow blew up, log out
        .then(
          // refresh token request succeeded
          () => {
            refreshing = false;
            return dispatch(sparkpostRequest(action));
          },
          // refresh token request failed
          (err) => {
            dispatch(logout());
            throw err;
          }
        );
    }

    // If we have a 403 or a 401 and we're not refreshing, log the user out silently
    if (response.status === 401 || response.status === 403) {
      dispatch(logout());
      throw err;
    }

    // any other API error should automatically fail, to be handled in the reducers/components
    dispatch({
      type: types.FAIL,
      payload: { message, response },
      meta
    });

    if (response.status >= 500) {
      dispatch(showAlert({ type: 'error', message: 'Something went wrong.', details: message }));
    }

    throw err;
  }
});

export default (action) => (dispatch, getState) => {
  // check for refreshing and exit early, otherwise call factory-made function
  if (refreshing) {
    return redispatchAfterRefresh(action, dispatch);
  }

  return dispatch(sparkpostRequest(action));
};
