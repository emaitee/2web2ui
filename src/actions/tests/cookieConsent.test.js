import { giveConsent } from '../cookieConsent';

import helpers from 'src/helpers/cookieConsent';

jest.mock('src/helpers/cookieConsent');

describe('Action Creator: cookie consent ', () => {
  describe('giveConsent', () => {
    it('should set the cookie', () => {
      helpers.setCookie.mockReturnValue(true);
      expect(giveConsent()).toMatchSnapshot();
      expect(helpers.setCookie).toHaveBeenCalledTimes(1);
    });
  });
});

