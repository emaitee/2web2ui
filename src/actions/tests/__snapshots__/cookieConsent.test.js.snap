// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Creator: cookie consent  giveConsent should set the cookie 1`] = `
Object {
  "type": "GIVE_COOKIE_CONSENT",
}
`;
